File: Farmsy/Features/Detail/FarmProductsSection.swift
import SwiftUI

/// What this farm sells, as tappable chips, and the way into the Shopping tab.
///
/// The chips are the list items whose served terms appear in the farm's produce
/// text (curated first, inferred as a fallback, same rule as the web). Tapping
/// one adds it to — or removes it from — the shopping list, so "add eggs to
/// shopping" is one tap from any farm card. Inferred produce is labelled,
/// because a language model read it out of the description and nobody confirmed it.
struct FarmProductsSection: View {
    let pin: FarmPin
    let detail: FarmDetail?

    @Environment(FarmsStore.self) private var farms
    @Environment(TripStore.self) private var trip
    @Environment(\.shell) private var shell
    /// Dismisses the farm card itself — this section lives inside that sheet,
    /// and AppShell's binding clears `selectedPin` when it closes.
    @Environment(\.dismiss) private var dismiss
    @State private var catalogue = ShoppingItems.shared

    private var sells: String? {
        detail?.displayProduce ?? farms.produceByOsm[pin.osmId]
    }

    private var inferred: Bool {
        (detail?.produce?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true)
            && detail?.produceInferred != nil
    }

    private var items: [ShoppingItem] {
        guard let sells else { return [] }
        return catalogue.items.filter { ProductMatch.covers(sells, terms: $0.terms) }
    }

    private var allOnList: Bool {
        !items.isEmpty && items.allSatisfy { trip.wantedProducts.contains($0.id) }
    }

    var body: some View {
        if !items.isEmpty {
            VStack(alignment: .leading, spacing: Space.s3) {
                HStack(alignment: .firstTextBaseline) {
                    Text("Products").font(.ui(17, .bold)).foregroundStyle(Color.ink)
                    if inferred {
                        Text("likely").role(.caption, .inkFaint)
                    }
                    Spacer()
                    Button(allOnList ? String(localized: "On your list") : String(localized: "Add all to shopping")) {
                        Haptics.success()
                        for item in items where !trip.wantedProducts.contains(item.id) {
                            trip.toggleProduct(item.id)
                        }
                        // Close the card before switching tabs: it is a sheet
                        // over the map, so switching underneath it left the
                        // Shopping tab hidden behind an open farm card.
                        if allOnList { dismiss(); shell.showTab(.shopping) }
                    }
                    .buttonStyle(PillButtonStyle(.text, size: .small))
                }
                FlowRow(spacing: Space.s2) {
                    ForEach(items) { item in
                        Chip(label: item.label, emoji: item.emoji,
                             selected: trip.wantedProducts.contains(item.id)) {
                            trip.toggleProduct(item.id)
                        }
                    }
                }
                Text("Tap a product to put it on your shopping list.")
                    .role(.caption, .inkMuted)
            }
            .task { await catalogue.loadIfNeeded() }
            .task { await farms.loadFlagsIfNeeded() }
        }
    }
}

/// "Report incorrect information": six reasons and an optional note, sent as a
/// contact message so it lands where Neil already reads (contact_submissions,
/// the activity log, the admin inbox email). No new table until the volume
/// justifies one.
struct ReportInfoSheet: View {
    let pin: FarmPin

    @Environment(\.dismiss) private var dismiss
    @Environment(SessionStore.self) private var session

    enum Reason: String, CaseIterable, Identifiable {
        case closed, hours, products, location, duplicate, other
        var id: String { rawValue }
        var label: String {
            switch self {
            case .closed: String(localized: "Closed permanently")
            case .hours: String(localized: "Opening hours are wrong")
            case .products: String(localized: "Products are wrong")
            case .location: String(localized: "Wrong location")
            case .duplicate: String(localized: "Duplicate farm")
            case .other: String(localized: "Something else")
            }
        }
    }

    @State private var reason: Reason?
    @State private var note = ""
    @State private var sending = false
    @State private var failed = false
    @State private var sent = false

    var body: some View {
        VStack(spacing: 0) {
            ScreenHeader(String(localized: "Report incorrect info"), compact: true) {
                IconButton("xmark", label: String(localized: "Close"), small: true) { dismiss() }
            }
            .padding(.top, Space.s2)
            ScrollView(showsIndicators: false) {
                VStack(alignment: .leading, spacing: Space.s4) {
                    if sent {
                        EmptyState(icon: "checkmark.circle",
                                   title: String(localized: "Thank you"),
                                   text: String(localized: "We check every report by hand and fix the listing."),
                                   action: (String(localized: "Done"), { dismiss() }))
                    } else {
                        Text(pin.name).role(.heading)
                        Text("What is wrong?").role(.bodySm, .inkMuted)
                        RowGroup {
                            ForEach(Reason.allCases) { r in
                                Row(icon: reason == r ? "checkmark.circle.fill" : "circle",
                                    title: r.label, chevron: false) { reason = r }
                            }
                        }
                        TextField(String(localized: "Anything else we should know? (optional)"), text: $note, axis: .vertical)
                            .font(.ui(16))
                            .lineLimit(3...6)
                            .padding(Space.s4)
                            .background(Color.surface, in: RoundedRectangle(cornerRadius: Radius.input, style: .continuous))
                            .overlay(RoundedRectangle(cornerRadius: Radius.input, style: .continuous).stroke(Color.hairline, lineWidth: 1))
                        if failed {
                            Text("Couldn't send. Check your connection and try again.").role(.caption, .critical)
                        }
                        Button {
                            Task { await send() }
                        } label: {
                            if sending { ProgressView().tint(.white) } else { Text("Send report") }
                        }
                        .buttonStyle(PillButtonStyle(.primary, size: .large, block: true))
                        .disabled(reason == nil || sending)
                    }
                }
                .padding(.horizontal, Space.s4)
                .padding(.vertical, Space.s4)
            }
        }
        .background(Color.cream.ignoresSafeArea())
    }

    private func send() async {
        guard let reason else { return }
        sending = true
        failed = false
        var request = URLRequest(url: Backend.webAPI.appending(path: "contact"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body: [String: String] = [
            "name": session.displayName,
            "email": session.email.isEmpty ? "anonymous@farmsy.app" : session.email,
            "topic": "Farm correction: \(reason.rawValue)",
            "message": "\(pin.name) (osm \(pin.osmId), \(pin.city ?? "-"))\nReason: \(reason.label)\n\(note)",
            "source": "ios_app",
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        let ok = (try? await URLSession.shared.data(for: request))
            .flatMap { ($0.1 as? HTTPURLResponse)?.statusCode }
            .map { (200..<300).contains($0) } ?? false
        sending = false
        if ok {
            Haptics.success()
            sent = true
        } else {
            failed = true
        }
    }
}

File: android/app/src/main/java/app/farmsy/android/features/detail/FarmProductsSection.kt
package app.farmsy.android.features.detail

import android.os.Build
import android.view.HapticFeedbackConstants
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.outlined.CheckCircle
import androidx.compose.material.icons.outlined.Circle
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalView
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.farmsy.android.LocalFarms
import app.farmsy.android.LocalSession
import app.farmsy.android.LocalTrip
import app.farmsy.android.R
import app.farmsy.android.core.Backend
import app.farmsy.android.core.FarmDetail
import app.farmsy.android.core.FarmPin
import app.farmsy.android.core.ProductMatch
import app.farmsy.android.core.ShoppingItems
import app.farmsy.android.core.httpClient
import app.farmsy.android.features.main.AppTab
import app.farmsy.android.features.main.LocalShell
import app.farmsy.android.ui.theme.Chip
import app.farmsy.android.ui.theme.EmptyState
import app.farmsy.android.ui.theme.FarmsyColors
import app.farmsy.android.ui.theme.IconButton
import app.farmsy.android.ui.theme.ListRow
import app.farmsy.android.ui.theme.PillButton
import app.farmsy.android.ui.theme.PillSize
import app.farmsy.android.ui.theme.PillVariant
import app.farmsy.android.ui.theme.Radius
import app.farmsy.android.ui.theme.RowGroup
import app.farmsy.android.ui.theme.ScreenHeader
import app.farmsy.android.ui.theme.Space
import app.farmsy.android.ui.theme.TextRole
import app.farmsy.android.ui.theme.role
import app.farmsy.android.ui.theme.ui
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType
import kotlinx.coroutines.launch
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/// What this farm sells, as tappable chips, and the way into the Shopping tab.
/// Mirrors iOS FarmProductsSection.
///
/// The chips are the list items whose served terms appear in the farm's produce
/// text (curated first, inferred as a fallback, same rule as the web). Tapping
/// one adds it to — or removes it from — the shopping list. Inferred produce is
/// labelled, because a language model read it out of the description and nobody
/// confirmed it.
@OptIn(ExperimentalLayoutApi::class)
@Composable
fun FarmProductsSection(pin: FarmPin, detail: FarmDetail?, onClose: () -> Unit) {
    val farms = LocalFarms.current
    val trip = LocalTrip.current
    val shell = LocalShell.current
    val context = LocalContext.current
    val catalogue by ShoppingItems.items.collectAsState()
    val wanted by trip.wantedProducts.collectAsState()
    // produceByOsm is a plain map; this flag is what tells us it has arrived.
    val flagsLoaded by farms.flagsLoaded.collectAsState()
    val language = remember { ShoppingItems.language(context) }

    LaunchedEffect(Unit) { ShoppingItems.loadIfNeeded(); farms.loadFlagsIfNeeded() }

    val sells = detail?.displayProduce ?: if (flagsLoaded) farms.produceByOsm[pin.osmId] else null
    val inferred = detail?.produce.isNullOrBlank() && detail?.produceInferred != null
    val items = remember(sells, catalogue) {
        if (sells == null) emptyList() else catalogue.filter { ProductMatch.covers(sells, it.terms) }
    }
    if (items.isEmpty()) return
    val allOnList = items.all { it.id in wanted }

    Column(Modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(Space.s3)) {
        Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(Space.s2)) {
            Text(stringResource(R.string.products), style = ui(17.sp, FontWeight.Bold), color = FarmsyColors.ink)
            if (inferred) Text(stringResource(R.string.products_likely), style = role(TextRole.CAPTION), color = FarmsyColors.inkFaint)
            Spacer(Modifier.weight(1f))
            PillButton(
                stringResource(if (allOnList) R.string.on_your_list else R.string.add_all_to_shopping),
                PillVariant.TEXT, PillSize.SMALL,
            ) {
                // Close the card before switching tabs: it is a sheet over the
                // map, so switching underneath it left the Shopping tab hidden
                // behind an open farm card (final review #7, iOS twin).
                if (allOnList) { onClose(); shell.showTab(AppTab.SHOPPING) }
                else items.filter { it.id !in wanted }.forEach { trip.toggleProduct(it.id) }
            }
        }
        FlowRow(horizontalArrangement = Arrangement.spacedBy(Space.s2), verticalArrangement = Arrangement.spacedBy(Space.s2)) {
            items.forEach { item ->
                Chip(item.label(language), emoji = item.emoji, selected = item.id in wanted) { trip.toggleProduct(item.id) }
            }
        }
        Text(stringResource(R.string.products_tap_hint), style = role(TextRole.CAPTION), color = FarmsyColors.inkMuted)
    }
}

/// "Report incorrect information": six reasons and an optional note, sent as a
/// contact message so it lands where Neil already reads (contact_submissions,
/// the activity log, the admin inbox email). No new table until the volume
/// justifies one. Mirrors iOS ReportInfoSheet.
private enum class ReportReason(val id: String, val labelRes: Int) {
    CLOSED("closed", R.string.report_reason_closed),
    HOURS("hours", R.string.report_reason_hours),
    PRODUCTS("products", R.string.report_reason_products),
    LOCATION("location", R.string.report_reason_location),
    DUPLICATE("duplicate", R.string.report_reason_duplicate),
    OTHER("other", R.string.report_reason_other),
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ReportInfoSheet(pin: FarmPin, onDismiss: () -> Unit) {
    val session = LocalSession.current
    val context = LocalContext.current
    val view = LocalView.current
    val scope = rememberCoroutineScope()
    var reason by remember { mutableStateOf<ReportReason?>(null) }
    var note by remember { mutableStateOf("") }
    var sending by remember { mutableStateOf(false) }
    var failed by remember { mutableStateOf(false) }
    var sent by remember { mutableStateOf(false) }

    fun send() {
        val r = reason ?: return
        scope.launch {
            sending = true; failed = false
            val body = buildJsonObject {
                put("name", session.displayName)
                put("email", session.email.ifEmpty { "anonymous@farmsy.app" })
                put("topic", "Farm correction: ${r.id}")
                put("message", "${pin.name} (osm ${pin.osmId}, ${pin.city ?: "-"})\nReason: ${context.getString(r.labelRes)}\n$note")
                put("source", "android_app")
            }
            val ok = runCatching {
                httpClient.post("${Backend.WEB_API}/contact") {
                    contentType(ContentType.Application.Json)
                    setBody(body.toString())
                }.status.value in 200..299
            }.getOrDefault(false)
            sending = false
            if (ok) {
                if (app.farmsy.android.ui.theme.Haptics.enabled) view.performHapticFeedback(if (Build.VERSION.SDK_INT >= 30) HapticFeedbackConstants.CONFIRM else HapticFeedbackConstants.VIRTUAL_KEY)
                sent = true
            } else failed = true
        }
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
        containerColor = FarmsyColors.cream,
    ) {
        ScreenHeader(stringResource(R.string.report_incorrect_info), compact = true) {
            IconButton(Icons.Filled.Close, label = stringResource(R.string.close), small = true, onClick = onDismiss)
        }
        Column(
            Modifier.fillMaxWidth().verticalScroll(rememberScrollState()).padding(Space.s4),
            verticalArrangement = Arrangement.spacedBy(Space.s4),
        ) {
            if (sent) {
                EmptyState(
                    Icons.Outlined.CheckCircle, stringResource(R.string.thank_you), stringResource(R.string.report_thank_you_body),
                    action = stringResource(R.string.done) to onDismiss,
                )
            } else {
                Text(pin.name, style = role(TextRole.HEADING), color = FarmsyColors.ink)
                Text(stringResource(R.string.report_what_is_wrong), style = role(TextRole.BODY_SM), color = FarmsyColors.inkMuted)
                RowGroup {
                    ReportReason.entries.forEach { r ->
                        ListRow(
                            icon = if (reason == r) Icons.Filled.CheckCircle else Icons.Outlined.Circle,
                            title = stringResource(r.labelRes), chevron = false,
                        ) { reason = r }
                    }
                }
                val shape = RoundedCornerShape(Radius.input)
                BasicTextField(
                    value = note, onValueChange = { note = it }, minLines = 3, maxLines = 6,
                    textStyle = ui(16.sp).copy(color = FarmsyColors.ink),
                    modifier = Modifier.fillMaxWidth().background(FarmsyColors.surface, shape).border(1.dp, FarmsyColors.hairline, shape).padding(Space.s4),
                    decorationBox = { inner ->
                        if (note.isEmpty()) Text(stringResource(R.string.report_note_placeholder), style = ui(16.sp), color = FarmsyColors.inkMuted)
                        inner()
                    },
                )
                if (failed) Text(stringResource(R.string.report_send_failed), style = role(TextRole.CAPTION), color = FarmsyColors.critical)
                PillButton(stringResource(R.string.report_send), PillVariant.PRIMARY, PillSize.LARGE, block = true, enabled = reason != null && !sending, onClick = ::send)
            }
        }
    }
}
